align/align.go

```diff
@@ -91,6 +82,74 @@ func NeedlemanWunsch(stringA string, stringB string, scoring Scoring) (int, stri
 	return matrix[columnLengthM][rowLengthN], string(alignA), string(alignB)
 }
 
+// SmithWaterman
+func SmithWaterman(stringA string, stringB string, scoring Scoring) (int, string, string) {
+
+	columnLengthM, rowLengthN := len(stringA), len(stringB)
+
+	// Initialize the alignment matrix
+	matrix := make([][]int, columnLengthM+1)
+	for i := 0; i <= columnLengthM; i++ {
+		matrix[i] = make([]int, rowLengthN+1)
+	}
+
+	// Initialize variables to keep track of the maximum score and its position
+	maxScore := 0
+	maxScoreRow := 0
+	maxScoreCol := 0
+
+	// Fill the alignment matrix
+	for i := 1; i <= columnLengthM; i++ {
+		for j := 1; j <= rowLengthN; j++ {
+			diagScore := matrix[i-1][j-1] + score(stringA[i-1], stringB[j-1], scoring)
+			upScore := matrix[i-1][j] + scoring.GapPenalty
+			leftScore := matrix[i][j-1] + scoring.GapPenalty
+			matrix[i][j] = max(0, max(diagScore, max(upScore, leftScore)))
+
+			if matrix[i][j] > maxScore {
+				maxScore = matrix[i][j]
+				maxScoreRow = i
+				maxScoreCol = j
+			}
+		}
+	}
+
+	// Traceback to construct the aligned strings
+	alignA := ""
+	alignB := ""
+	i := maxScoreRow
+	j := maxScoreCol
+	for matrix[i][j] > 0 {
+		if matrix[i][j] == matrix[i-1][j-1]+score(stringA[i-1], stringB[j-1], scoring) {
+			alignA = string(stringA[i-1]) + alignA
+			alignB = string(stringB[j-1]) + alignB
+			i--
+			j--
+		} else if matrix[i][j] == matrix[i-1][j]+scoring.GapPenalty {
+			alignA = string(stringA[i-1]) + alignA
+			alignB = "-" + alignB
+			i--
+		} else if matrix[i][j] == matrix[i][j-1]+scoring.GapPenalty {
+			alignA = "-" + alignA
+			alignB = string(stringB[j-1]) + alignB
+			j--
+		} else {
```

Comment:
is there any case where this can happen? Shouldn't panic but should return an error if this case is possible.

Reply:
I am pretty sure this case should never happen. It would mean that the matrix was constructed erroneously or there is a bug somewhere.